### PR TITLE
DOCS-1006: Viam Visual Odometry module

### PR DESCRIPTION
**Pull request by andf-viam**
Add docs for the new https://github.com/viamrobotics/viam-visual-odometry module repo.

- Included cool gif from upstream repo in doc, but could remove vid controls depending. I like that I can watch at command, and pause at command. Fast-moving video on careful-reading reference page gets at me. However, the vid itself is excellent in conveying the content of this page. Left vid controls in as a compromise, but aware of other opinions.
- Included some duplicate content from calibrate camera page, because this module requires some JSON object reshuffling in the copy&pate transfer. Opted for explicitness here.
- EDIT: Now with link to module page in Registry, and with card under Module Resources > Examples.

#### Staging

- [/components/movement-sensor/viam-visual-odometry/](https://docs-test.viam.dev/ae972795178d36723f10bfd3dd6e26040603ac66/public/components/movement-sensor/viam-visual-odometry/)

### CODE REVIEW

**Comment by andf-viam** [comment]
Hi @Rob1in , have you had a chance to review these docs? We'd like to get this merged by the Registry launch, so we can link to it from modular resource walkthroughs. Thank you!

**Comment by Rob1in** [comment]
Hi @andf-viam, sorry I completely missed that, doing it right now!

File: docs/components/movement-sensor/viam-visual-odometry.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      "type": "movement_sensor",
+      "namespace": "rdk",
+      "model": "viam:visual_odometry:opencv_orb",
+      "attributes": {
+        "camera_name": "my-camera", 
+        "time_between_frames_s": 0.2, 
+        "lowe_ratio_threshold": 0.75
+      },
+      "depends_on": [
+        "my-camera"
```
Comment (Rob1in) [inline docs/components/movement-sensor/viam-visual-odometry.md:185]: It should work with the `depends_on` field  empty. An error will be logged if the attribute `camera_name` is empty.

File: docs/components/movement-sensor/viam-visual-odometry.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+|`orb_scale_factor`| float | Optional | `1.2` | Pyramid decimation ratio, greater than 1. |
+|`orb_WTA_K`| int | Optional | `2` | Number of points that produce each element of the oriented BRIEF descriptor. |
+|`matcher`| string | Optional | `"flann"` | Either `"flann"` for [FLANN based matcher](https://docs.opencv.org/3.4/d5/d6f/tutorial_feature_flann_matcher.html) or `"BF"` for brute force matcher. The FLANN matcher will look for the two best matches using the KNN method so Lowe's ratio test can be performed afterward. The [brute force matcher](https://docs.opencv.org/4.x/dc/dc3/tutorial_py_matcher.html) uses Hamming norm. |
+|`lowe_ratio_threshold`| float | Optional | `0.8` | Threshold value to check if the best match is significantly better than the second best match. This value will not be used if brute force matcher is chosen. |
+| `ransac_prob` | float | Optional | `0.99` | Probability to find a subset without outliers in it. Defines the number of iterations to filter the outliers. The number of iterations is roughly given by `$k = \frac{\log(1-p)}{\log(1-w^n)}$`, where `$n$` is the number of points and `$w$` is the ratio of inliers to total points.|
```
Comment (Rob1in) [inline docs/components/movement-sensor/viam-visual-odometry.md:250]: Might be nothing but when I built the doc with `make serve-dev-draft` the LateX formula for the number of iterations didn't render well like in [here](https://github.com/viamrobotics/viam-visual-odometry#attributes-description).

File: docs/components/movement-sensor/viam-visual-odometry.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+      "type": "movement_sensor",    
+      "namespace": "rdk",
+      "model": "viam:visual_odometry:opencv_orb",
+      "attributes": {
+        "camera_name": "<your-camera-name>", 
+        "time_between_frames_s": <time_seconds>, 
+        "lowe_ratio_threshold": <lowe_ratio_threshold>
+      },
+      "depends_on": [
+        "<your-camera-name>"
```
Comment (Rob1in) [inline docs/components/movement-sensor/viam-visual-odometry.md:130]: It should work with the depends_on field empty. An error will be logged if the attribute camera_name is empty.

**Review by Rob1in — APPROVED** [review]
Mostly lgtm, the main thing is the sensor fusion aspect. Most cheap dedicated movement sensors would perform the same/worse than monocular visual odometry, that's why I would recommend using two movement sensors to users who want robust measurement.

File: docs/components/movement-sensor/viam-visual-odometry.md (hunk 2 above)
Comment (andf-viam) [reply docs/components/movement-sensor/viam-visual-odometry.md:250]: Yes, I believe this is a dev build limitation. I will make sure it renders for prod publish (it should with existing tooling and default settings, but will double-check!). Appreciate the callout!

File: docs/components/movement-sensor/viam-visual-odometry.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
+---
+title: "Configure Visual Odometry"
+linkTitle: "viam-visual-odometry"
+weight: 40
+type: "docs"
+description: "Configure viam-visual-odometry, a modular resource that derives movement data from a camera."
+images: ["/icons/components/imu.svg"]
```
Comment (npentrel) [inline docs/components/movement-sensor/viam-visual-odometry.md:7]:
```suggestion
images: ["/icons/components/imu.svg"]
katex: true
```

File: docs/components/movement-sensor/viam-visual-odometry.md (hunk 2 above)
Comment (npentrel) [reply docs/components/movement-sensor/viam-visual-odometry.md:249]: I think you need to enable katex - haven't tried though. and I don't think you need backticks then
```suggestion
| `ransac_prob` | float | Optional | `0.99` | Probability to find a subset without outliers in it. Defines the number of iterations to filter the outliers. The number of iterations is roughly given by $k = \frac{\log(1-p)}{\log(1-w^n)}$, where `$n$` is the number of points and `$w$` is the ratio of inliers to total points.|
```

**Review by npentrel — APPROVED** [review]
LGTM % latex issue

File: docs/components/movement-sensor/viam-visual-odometry.md (hunk 4 above)
Comment (andf-viam) [reply docs/components/movement-sensor/viam-visual-odometry.md:7]: oh! TIL! Thank you!

**Pull request merged**